### ISSUE TITLE
Performance Issues: Slow Response, Requires Frequent Restarts, Affects TS Files

### ISSUE THREAD

**Issue by bmitchenko** [angular/vscode-ng-language-service#2131]
### Is this a regression?

Unknown, but recent updates seem to have worsened the issue.

### Description

The Angular Language Service in VSCode is significantly slowing down development. The extension often becomes unresponsive, leading to delays in template and TypeScript file analysis. Restarting the extension temporarily resolves the issue but hampers workflow. These problems are consistent with the experiences shared by users in the [VSCode Marketplace reviews](https://marketplace.visualstudio.com/items?itemName=Angular.ng-template&ssr=false#review-details).

Key Issues:

- Extremely slow response time when editing Angular templates or TypeScript files.
- The extension requires frequent restarts to restore functionality.
- Development performance is degraded even when working on non-template (TS) files.

## Bug Type
What does this bug affect

- [x] Angular Language Service VSCode extension
- [ ] Angular Language Service server

## Reproduction

1. Open an Angular workspace in VSCode with the Angular Language Service enabled.
2. Begin editing a template or TypeScript file.
3. Observe the increasing latency in language features (e.g., intellisense, error highlighting).
4. Notice that restarting the extension temporarily improves performance, but the issue recurs.

**Expected behavior**

The Angular Language Service should provide quick and responsive language features without requiring frequent restarts or causing performance issues.

## 🌍  Your Environment
**Angular Version:**
<pre><code>Angular: 19.0.1
Angular CLI: 19.0.2
Node: 22.9.0
Package Manager: npm 10.8.3
OS: win32 x64</code></pre>

Extension Version:
<pre><code>19.0.3</code></pre>

VSCode Version:
<pre><code>Version: 1.96.1 (user setup)
Commit: 42b266171e51a016313f47d0c48aca9295b9cbb2
Date: 2024-12-17T17:50:05.206Z
Electron: 32.2.6
ElectronBuildId: 10629634
Chromium: 128.0.6613.186
Node.js: 20.18.1
V8: 12.8.374.38-electron.0
OS: Windows_NT x64 10.0.26100</code></pre>

Operating System:
<pre><code>Windows 11 Home 24H2 OS Build 26100.2314</code></pre>

Extension options:
<pre><code>Default settings.</code></pre>

**Comment by glued** [comment]
seeing this too ever since updating to Angular 19

**Comment by ShacharHarshuv** [comment]
I'm not sure if it's the same issue, but I want to contribute my experience - 
- As far as I can see it only affects "find references" / "go to definition" and similar. Intelligence seems to be working fine, but I might be missing cases where it's not.
- Disabling the Angular Language Service makes typescript functionality work fine (no excessive delays) but obviously it's not a solution because I won't get language support for html.
- It's reproducing in every extension version I checked, with my project being in Angular 19.
- I checked out an older version of my project that used Angular 18, the issue is still reproducing, even if I install the matching version of the extension. 

This is right now the only reason I don't use vscode. Would love some updates if this is something that is going to be fixed so I can try vscode again.

**Comment by Canolz** [comment]
cmd/ctrl clicking on an import and getting a result takes 5-15 seconds and depending if it's in the node modules it takes even longer. Specs are: M2 Macbook pro max 64gb

**Comment by atscott** [comment]
In the past for performance debugging, we have often used https://github.com/atscott/ng-speed-rebuild. I haven't observed performance issues that have previously come up when something is awry. 

To those experiencing issues:

* Are you still seeing them
* Can you provide a reproduction repository
* Is it reproducible in https://github.com/atscott/ng-speed-rebuild
* Are you also seeing slow builds for the application? This could be a duplicate of https://github.com/angular/angular/issues/59040 in that case.

**Comment by atscott** [comment]
> Update: I moved to zed and am using the angular LSP on that. My performance issues have all disappeared!

Thats a really interesting datapoint. Since the extension uses the language server npm package that we publish here, there’s not much difference in how these should perform other than VSCode vs Zed as an IDE. There’s very little we do on the client side for the VSCode extension other than short-circuit prevention from even sending requests to the server when not in an Angular context.

**Comment by bmitchenko** [comment]
I'd like to point out that the issue is not with the build process, but with intellisense in VS Code. After disabling the Angular Language Service plugin, intellisense starts working instantly. With the plugin enabled, there are significant delays in autocompletion and other intellisense features.

This problem specifically occurs in large repositories, while performance is normal in smaller ones. This makes it difficult to create a simple reproduction case.

**Comment by atscott** [comment]
> I'd like to point out that the issue is not with the build process, but with intellisense in VS Code

This is one and the same. The language service needs to compile the whole program in order to provide autocomplete features. NgModules require global knowledge. Standalone components can make this better and there is work necessary to improve partial compilation in this area. 

Autocomplete is disabled in TypeScript contexts. It should not be affecting anything other than autocomplete inside templates. Incremental compilation for changes to template files should be reasonably fast.

**Comment by ShacharHarshuv** [comment]
I tried again now with both my repo and the example repo and it seems to work much better. There is a noticeable delay in the first "Go to action" but I've tried a bunch after it and the rest are smooth, which mean I can work. 
I'm not sure what got updated though that fixed it (vscode / LS / windows..?)
I'll update if I see that happening again!

**Comment by atscott** [comment]
Given that many folks have reported not being able to reproduce this anymore, I'm going to close this rather than keeping it as a catch-all for any performance issues that might happen in the future, which may have different causes. If problems come up again, please feel free to open a new issue.

**Comment by ShacharHarshuv** [comment]
Btw, I recently tried Cursor AI (a vscode fork) and it still reproduces there. Not sure if it's because they don't have some sort of update that is missing. Hopefully it will resolve on its own... but I'm really not sure what the cause was.

**Comment by atscott** [comment]
> Btw, I recently tried Cursor AI (a vscode fork) and it still reproduces there

Based on your previous comment, I’m assuming “it” is the very first request while subsequent requests are fine. This is entirely expected. Angular needs to compile the program to provide its information. Subsequent requests should be more performant with incremental rebuilds.

**Comment by bmitchenko** [comment]
I'd like to note that the problem still persists. It can be consistently reproduced in many large repositories, though creating an isolated reproduction case has proven difficult due to the nature of the issue.

I don't see 'many reports' in this thread claiming inability to reproduce the issue - quite the opposite. The original problem has received numerous likes, and there are negative reviews in the marketplace reflecting these ongoing concerns. The closure of this issue seems premature as the underlying performance problems continue to affect users.

**Comment by bmitchenko** [comment]
Go To Definition, IntelliSense, and other actions continue to work very slowly. Reloading the Angular Language Service makes these actions work instantly with no waiting time. However, after some time, they start lagging again until the next restart of Angular Language Service. If Angular Language Service is completely disabled, there are no performance issues at all.

**Comment by Canolz** [comment]
> Go To Definition, IntelliSense, and other actions continue to work very slowly. Reloading the Angular Language Service makes these actions work instantly with no waiting time. However, after some time, they start lagging again until the next restart of Angular Language Service. If Angular Language Service is completely disabled, there are no performance issues at all.

Exactly. I have to restart the extension multiple times in my workday if I feel I have to wait too long. 

Please reopen this issue

**Comment by atscott** [comment]
Thank you for your responses. Re-opening. Please note that there is still currently not enough information for this to be actionable.

* can you reproduce the issue in https://github.com/atscott/ng-speed-rebuild
* can you provide some of the trace logs when the issue begins to reproduce?
* could run a [profile](https://github.com/microsoft/vscode/wiki/performance-issues#profile-the-running-extensions) when the issue starts to reproduce and provide it here? A before and after restarting the extension could be useful (and just want to double check: this is restarting the _extension_ only and not all of vscode?)
* additional information about the vscode process could be useful as well, such as https://github.com/microsoft/vscode/wiki/performance-issues#collect-information-about-the-running-vs-code or other profiling such as the render or shared process as mentioned on that page, depending on whether those are problematic. Again, a before and after restarting the extension could be useful. 
* what actions, specifically, are slow? The issue mentions autocomplete in typescript files, outside of Angular contexts (so not in host bindings or inline templates). This would be very interesting because the extension isn’t doing anything here except checking the cursor position and then returning without doing anything else.

**Comment by ShacharHarshuv** [comment]
> Based on your previous comment, I’m assuming “it” is the very first request while subsequent requests are fine. 

No, I was actually referring to my first description of the issue: https://github.com/angular/vscode-ng-language-service/issues/2131#issuecomment-2719113657

> Reloading the Angular Language Service makes these actions work instantly with no waiting time.

I don't experience this. To me it's slow from the beginning. However, I can confirm that disabling the Angular Language Service resolves all of these issues (i.e. the TS Language Service works. Obviously you lose intellisense for templates)

@atscott I'll try these things later. The one question I can answer: specifically the "go to" action is slow. I haven't had issues with autocomplete or intelligence. It might be that my issue is different from others described here.

**Comment by atscott** [comment]
> specifically the "go to" action is slow

Does this include "go to definition" and "go to type definition"? Excluding when you perform those actions when you're already on the definition, which VSCode changes to "go to references" instead. References may include references in templates so that would also require compiling the Angular program.

**Comment by atscott** [comment]
In the absence of more information, my current plan is to add an option in the extension settings to disable the references and rename functionality and close this issue. 

References and renaming will always require global compilation to convert templates into their corresponding Typescript blocks used for type checking. 

Beyond that, I don’t see anything actionable on the language service extension side of things. Further performance improvements could likely be made in certain areas related to incremental/partial compilations with authoring format features such standalone and selectorless.

**Comment by ShacharHarshuv** [comment]
Hey @atscott, sorry for taking so long to reply, I was busy with work stuff.
I noticed now that I'm unable to consistently reproduce this, but it does still happen. Upon further examination, it appears it does happen after using the server for a while, and "resolved" when restarting vscode. This makes it difficult to recreate with the demo project you added. To makes things more difficult - once a performance issue occurs when one action "like go to definition" the same action with the same symbol works quickly, probably due to cache. 

I wonder if I will be able to share logs / profile when it does happen, even if I don't know how to purposefully make it happen. 
I'll try to do the profile thing you suggested, but I wanted to ask about "trace logs" which you mentioned - what are those and where can I find them?

**Comment by atscott** [comment]
@ShacharHarshuv server logs can be enabled and opened in the vscode command palette:

![Image](https://github.com/user-attachments/assets/97403ca4-c37b-409f-b696-7482d8cc22e1)

 The verbosity of logs can be configured in the extension settings: 

![Image](https://github.com/user-attachments/assets/b36ebeae-659a-4802-a058-78b510239120)

The trace between the language server and client is available in the output panel in VSCode: 

![Image](https://github.com/user-attachments/assets/81d9e42a-6ee3-43c1-858b-300007a828e1)

**Comment by atscott** [comment]
> You can see it took very long (>4s for a simple lookup) which is much more than it usually takes. (subsequent requests took 173ms).

Are the "subsequent requests" that are shorter only after you restarted the language server or the problem resolved itself immediately on its own for subsequent requests? I do totally agree that the 4s for a lookup after the program was already compiled is a problem, but much less so if it's a 1-time thing and resolves itself immediately.

**Comment by ShacharHarshuv** [comment]
> Are the "subsequent requests" that are shorter only after you restarted the language server or the problem resolved itself immediately on its own for subsequent requests?

Immediately on subsequent requests. I did not restart the server. It happens again "every once in a while" but frequent enough to being annoying.

Another possible hint - I tried to run the "running extensions" profiler while trying to do a look up that took too long, and the Angular Language Service actually shows pretty low:
Activation: 363ms
Profile: 0.56ms
(Can I extract more info from that?)

I wonder if it means the extension wasn't "working" but actually "waiting".  Like the "request" wasn't prioritized for processing or something.

**Comment by atscott** [comment]
Thanks. I can't think of anything off the top of my head for debugging this further right now, but it's a useful clue for some future digging.

**Comment by atscott** [comment]
Given the narrowed down problem space of references and rename requests, I was able to observe at least one issue relatively easily in https://github.com/atscott/ng-speed-rebuild. The first request for references takes a few seconds, as expected, since this requires global compilation. After changing _just_ a template file, a follow-up request for references _should_ be quick but was also slow. 

There was an issue with the incremental results not retaining those from the previous compilation. This was fixed in https://github.com/angular/angular/commit/c71adb1ad1bc67b18f586e5037221d744b76f51b and was included in the v20.0.1 release. I'm going to close this now as completed.

If the problem still exists, please [open a new issue](https://github.com/angular/vscode-ng-language-service/issues/new/choose) and follow the instructions in the issue template that include info on how to create a reproduction using our template.
